### PR TITLE
Implement compiler-supplied ids for java compilation

### PR DESCRIPTION
**Pull request by hegyibalint**
Our current approach to Java compilation diagnostic IDs and their respective label was a bit simple: we derived an ID and a message from the kind of diagnostic, which limited us to 5 categories overall.

The compiler provides us with separate keys for various problems, which we could use on our end to also create finer-grained reports.

Fixes [#30514](https://github.com/gradle/gradle/issues/30514)

### CODE REVIEW

File: platforms/jvm/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileProblemsIntegrationTest.groovy [hunk 1]
```diff
@@ -229,24 +244,28 @@ class JavaCompileProblemsIntegrationTest extends AbstractIntegrationSpec impleme
...
+            assertLocations(it, true)
+            severity == Severity.WARNING
+            fqid == 'compilation:java:compiler-warn-redundant-cast'
+            contextualLabel == 'redundant cast to java.lang.String'
         }
         verifyAll(receivedProblem(3)) {
-            assertProblem(it, "WARNING", true)
-            fqid == 'compilation:java:java-compilation-warning'
-            details == 'redundant cast to java.lang.String'
+            assertLocations(it, true)
```
Comment (donat) [inline platforms/jvm/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileProblemsIntegrationTest.groovy:265]: I was never a fan of boolean parameters for functions. Instead, It's not strictly related to this pr, but I'd recommend to introduce a new `assertPreciseLocation` method.

File: platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java [hunk 2]
```diff
@@ -159,26 +159,31 @@ private static void printDiagnosticCount(Log logger, String kind, int number) {
...
         spec.severity(mapKindToSeverity(diagnostic.getKind()));
+        addId(spec, diagnostic);
+        addContextualLabel(spec, diagnostic);
         addFormattedMessage(spec, diagnostic);
-        addDetails(spec, diagnostic);
         addLocations(spec, diagnostic);
     }
 
+    private static void addId(ProblemSpec spec, Diagnostic<? extends JavaFileObject> diagnostic) {
+        String kebabCode = diagnostic.getCode().replace('.', '-');
```
Comment (donat) [inline platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java:170]: This is a name property for the problem ID. A better name for the local variable would be `name` or `idName`.

File: platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java [hunk 3]
```diff
@@ -159,26 +159,31 @@ private static void printDiagnosticCount(Log logger, String kind, int number) {
...
+        spec.id(kebabCode, mapKindToLabel(diagnostic.getKind()), GradleCoreProblemGroup.compilation().java());
```
Comment (donat) [inline platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java:171]: mapKindToDisplayName

File: platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java [hunk 4]
```diff
@@ -242,16 +247,16 @@ private static String getPath(JavaFileObject fileObject) {
     private static String mapKindToId(Diagnostic.Kind kind) {
```
Comment (donat) [inline platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java:247]: Is this unused now?

File: testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/problems/KnownProblemIds.groovy [hunk 5]
```diff
@@ -58,14 +61,17 @@ class KnownProblemIds {
...
+     */
+    private static final HashMap<String, String> KNOWN_DEFINITIONS = [
         'problems-api:missing-id' : 'Problem id must be specified',
         'problems-api:unsupported-additional-data' : 'Unsupported additional data type',
         'compilation:groovy-dsl:compilation-failed' : 'Groovy DSL script compilation problem',
-        'compilation:java:java-compilation-error' : 'Java compilation error',
-        'compilation:java:java-compilation-failed' : 'Java compilation error',
-        'compilation:java:java-compilation-warning' : 'Java compilation warning',
-        'compilation:java:java-compilation-advice' : 'Java compilation note',
+        // Flexible category, as the category id's last component, and the message, will be supplied by the compiler
```
Comment (donat) [inline testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/problems/KnownProblemIds.groovy:73]: For the most part, I'd prefer exact content listed here. At the same time, if the list becomes unmanageable (dozens of tests fail and  the list of known IDs are unmanageable), then I'd switch to pattern matching. Is this the case here?

File: platforms/jvm/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileProblemsIntegrationTest.groovy (hunk 1 above)
Comment (hegyibalint) [reply platforms/jvm/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileProblemsIntegrationTest.groovy:266]: I would not force it into this PR, but we could make an issue to review the structure there.

File: platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java (hunk 2 above)
Comment (hegyibalint) [reply platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java]: Agreed, I've changed it!

File: platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java (hunk 3 above)
Comment (hegyibalint) [reply platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java]: Good catch, renamed!

File: platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java (hunk 4 above)
Comment (hegyibalint) [reply platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java]: Correct, removed!

File: platforms/jvm/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileProblemsIntegrationTest.groovy [hunk 6]
```diff
@@ -161,24 +167,28 @@ class JavaCompileProblemsIntegrationTest extends AbstractIntegrationSpec impleme
...
+            severity == Severity.WARNING
+            fqid == 'compilation:java:compiler-warn-redundant-cast'
```
Comment (wolfs) [inline platforms/jvm/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileProblemsIntegrationTest.groovy:190]: ❓ Does it make sense to have the severity in the id? And would it make sense to kick out the `compiler` prefix? So we'd end up with something like `compilation:java:redundant-cast`.

Comment (hegyibalint) [reply platforms/jvm/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileProblemsIntegrationTest.groovy:190]: IMO the ID is just an ID, and it's not an issue if it contains redundant information.
Even if I'm pretty certain that the prefix here is stable (e.g. JDK8's list of [compiler codes](https://github.com/openjdk/jdk/blob/fcbd7a77c40eec49b469a9a7d140678d5ed3076a/langtools/src/share/classes/com/sun/tools/javac/resources/compiler.properties)), I would still avoid making an assumption of the code format (apart from the dot separation).

You have something in mind, where the shorter ID might be beneficial?

File: testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/problems/KnownProblemIds.groovy (hunk 5 above)
Comment (hegyibalint) [reply testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/problems/KnownProblemIds.groovy]: I took a different approach here: I kept the literal message and only kept the end if IDs dynamic. I think this is a sweet spot for not making the list huge yet having confidence that the value incoming is a (probably) valid one.

File: platforms/jvm/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileProblemsIntegrationTest.groovy (hunk 6 above)
Comment (wolfs) [reply platforms/jvm/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileProblemsIntegrationTest.groovy:190]: Fine with me. Will some humans end up reading the ID at some point?

Comment (donat) [reply platforms/jvm/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileProblemsIntegrationTest.groovy:190]: In Buildship we will soon have a UI for problems. Here's a [screenshot](https://private-user-images.githubusercontent.com/419883/338255346-5597a534-9dc8-4122-92c7-a25af03b275c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg2NDA1NDcsIm5iZiI6MTcxODY0MDI0NywicGF0aCI6Ii80MTk4ODMvMzM4MjU1MzQ2LTU1OTdhNTM0LTlkYzgtNDEyMi05MmM3LWEyNWFmMDNiMjc1Yy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYxN1QxNjA0MDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yNTk2Y2VhNGE1ZDMwMGRkMTJmOTY1YjM5NWJlY2FjMGU2MmVkZjA0M2FhOGNlYWRkZDk4ZDRkYzliZGQzNWY5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.d9Kd9T1Q9pvAjAVMHohb5k4QBTUylE_vZ7x0zygifIM) from my [wip pr](https://github.com/eclipse/buildship/pull/1306).

File: testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/problems/KnownProblemIds.groovy [hunk 7]
```diff
@@ -63,15 +66,22 @@ class KnownProblemIds {
...
-        'compilation:java:java-compilation-error' : 'Java compilation error',
-        'compilation:java:java-compilation-failed' : 'Java compilation error',
-        'compilation:java:java-compilation-warning' : 'Java compilation warning',
-        'compilation:java:java-compilation-advice' : 'Java compilation note',
+        // Flexible java compilation categories
+        // The end of the category is matched with a regex, as there are many possible endings (and also changes with JDK versions)
+        // See compiler.java for the full list of diagnostic codes we use as categories (we replace the dots with dashes)
+        'compilation:java:compiler-err-.+' : 'Java compilation error',
+        'compilation:java:compiler-warn-.+' : 'Java compilation warning',
+        'compilation:java:compiler-note-.+' : 'Java compilation note',
```
Comment (donat) [inline testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/problems/KnownProblemIds.groovy:83]: This can probably go away.

Comment (hegyibalint) [reply testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/problems/KnownProblemIds.groovy:83]: I don't think so: wouldn't this check fail if we start emitting problems with these IDs?

Comment (donat) [reply testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/problems/KnownProblemIds.groovy:83]: ok

**Review by donat — APPROVED** [review]

**Pull request merged**